test/test_framework.py

```diff
@@ -321,6 +321,90 @@ def on_foo(self, event):
+
+        observed_events = []
+
+        class MyEvent(EventBase):
+            pass
+
+        class MyEvents(EventsBase):
+            foo = Event(MyEvent)
+
+        class MyNotifier(Object):
+            on = MyEvents()
+
+        class MyObserver(Object):
+            def __init__(self, framework, key):
+                super().__init__(framework, key)
+                self.key = key
+
+            def on_foo(self, event):
+                observed_events.append(f"foo-{self.key}")
+
+        pub = MyNotifier(framework, "1")
+        observers = []
+        for i in range(2, 10):
+            obs = MyObserver(framework, str(i))
+            observers.append(obs)
+            framework.observe(pub.on.foo, obs)
+            del obs
+        self.assertEqual(len(framework._observers), 8)
+        self.assertEqual(len(framework._observer), 8)
+        pub.on.foo.emit()
+        self.assertEqual(observed_events, ["foo-2", "foo-3", "foo-4", "foo-5", "foo-6", "foo-7", "foo-8", "foo-9"])
+        # Remove some, but not all, observers, see the remaining ones are still triggering
+        observers = observers[:2]
+        gc.collect()
+        pub.on.foo.emit()
+        self.assertEqual(observed_events, ["foo-2", "foo-3", "foo-4", "foo-5", "foo-6", "foo-7", "foo-8", "foo-9", "foo-2", "foo-3"])
+        self.assertEqual(len(framework._observer), 2)
+        self.assertEqual(len(framework._observers), 2)
+        # Delete the rest, and see that there are no new triggers
+        del observers
+        gc.collect()
+        pub.on.foo.emit()
+        self.assertEqual(observed_events, ["foo-2", "foo-3", "foo-4", "foo-5", "foo-6", "foo-7", "foo-8", "foo-9", "foo-2", "foo-3"])
+        self.assertEqual(len(framework._observer), 0)
+        self.assertEqual(len(framework._observers), 0)
+
+    def test_cleanup_and_reregister_observer(self):
+        framework = self.create_framework()
+
+        observed_events = []
+
+        class MyEvent(EventBase):
+            pass
+
+        class MyEvents(EventsBase):
+            foo = Event(MyEvent)
+
+        class MyNotifier(Object):
+            on = MyEvents()
+
+        class MyObserver(Object):
+            def __init__(self, framework, key):
+                super().__init__(framework, key)
+                self.key = key
+
+            def on_foo(self, event):
+                observed_events.append(f"foo-{self.key}")
+        pub = MyNotifier(framework, "1")
+        obs = MyObserver(framework, "2")
+        framework.observe(pub.on.foo, obs)
+        pub.on.foo.emit()
+        self.assertEqual(observed_events, ["foo-2"])
+        # Now, we delete the existing obs, which would cause it to be removed from observers,
+        # and then immediately register a new obs at the same path using a new object
+        del obs
+        new_obs = MyObserver(framework, "2")
+        framework.observe(pub.on.foo, new_obs)
+        pub.on.foo.emit()
+        self.assertEqual(observed_events, ["foo-2", "foo-2"])
```

Comment:
This is the test case that shows you can delete and reregister and still get the right events.

Reply:
Should this also include a test of the case where gc happens between the `del` and re-registration, just to confirm that case works as expected as well?

Reply:
Should this also include a test of the case where gc happens between the `del` and re-registration, just to confirm that case works as expected as well?